pkg/openstack/common.go

```diff
@@ -27,3 +41,255 @@ func EnsureDeleted(ctx context.Context, helper *helper.Helper, obj client.Object
+	namespace string,
+	labelSelectorMap map[string]string,
+) (*routev1.RouteList, error) {
+	routeList := &routev1.RouteList{}
+	listOpts := []client.ListOption{
+		client.InNamespace(namespace),
+		client.MatchingLabels(labelSelectorMap),
+	}
+
+	if err := h.GetClient().List(ctx, routeList, listOpts...); err != nil {
+		err = fmt.Errorf("Error listing routes for %s: %w", labelSelectorMap, err)
+		return nil, err
+	}
+
+	return routeList, nil
+}
+
+// EnsureRoute -
+func EnsureRoute(
+	ctx context.Context,
+	instance *corev1.OpenStackControlPlane,
+	helper *helper.Helper,
+	owner metav1.Object,
+	svcs *k8s_corev1.ServiceList,
+	svcOverrides map[string]service.OverrideSpec,
+	overrideSpec *route.OverrideSpec,
+	condType condition.Type,
+) (map[string]service.OverrideSpec, ctrl.Result, error) {
+
+	for _, svc := range svcs.Items {
+		rd := RouteDetails{
+			RouteName:         svc.Name,
+			Namespace:         svc.Namespace,
+			Endpoint:          svc.Annotations[service.AnnotationEndpointKey],
+			RouteOverrideSpec: overrideSpec,
+			ServiceSpec:       &svc,
+		}
+		svcOverride := svcOverrides[rd.Endpoint]
+
+		// If the service has the create ingress annotation and its a default ClusterIP service -> create route
+		if svc.ObjectMeta.Annotations[service.AnnotationIngressCreateKey] == "true" && svc.Spec.Type == k8s_corev1.ServiceTypeClusterIP {
+			instance.Status.Conditions.Set(condition.UnknownCondition(
+				condType,
+				condition.InitReason,
+				corev1.OpenStackControlPlaneExposeServiceReadyInitMessage,
+				owner.GetName(),
+				svc.Name,
+			))
+
+			if routeName, ok := svc.ObjectMeta.Annotations[service.AnnotationIngressNameKey]; ok {
+				rd.RouteName = routeName
+			}
+			if labelVal, ok := svcOverride.EmbeddedLabelsAnnotations.Labels[common.AppSelector]; ok {
+				rd.ServiceLabel = map[string]string{common.AppSelector: labelVal}
+			}
+
+			// check if there is already a route with common.AppSelector from the service
+			if svcLabelVal, ok := svc.Labels[common.AppSelector]; ok {
+				routes, err := GetRoutesListWithLabel(
+					ctx,
+					helper,
+					instance.Namespace,
+					map[string]string{common.AppSelector: svcLabelVal},
+				)
+				if err != nil {
+					return svcOverrides, ctrl.Result{}, err
+				}
+
+				// check the routes if name changed where we are the owner
+				for _, r := range routes.Items {
+					instanceRef := metav1.OwnerReference{
+						APIVersion:         instance.APIVersion,
+						Kind:               instance.Kind,
+						Name:               instance.GetName(),
+						UID:                instance.GetUID(),
+						BlockOwnerDeletion: pointer.Bool(true),
+						Controller:         pointer.Bool(true),
+					}
+
+					owner := metav1.GetControllerOf(&r.ObjectMeta)
+
+					if r.Name != rd.RouteName && owner != nil && referSameObject(*owner, instanceRef) {
+						// Delete any other owner refs from ref list to not block deletion until owners are gone
+						r.SetOwnerReferences([]metav1.OwnerReference{instanceRef})
```

Comment:
we add the service operator CR to the owner ref list. When the osctlplane CR gets deleted, this prevents that the route gets deleted immediately before the osp service resources are deleted. This is required that keystoneendpoints can be cleaned up properly without adding cleanup logic in the openstack-operator. In case of someone overrides the ingress_name after initial create, a new route gets created. In that case we override the refs to just the ctlplane CR ref.

Reply:
its gone now